AsyncDisplayKit/Details/ASCollectionDataController.mm

```diff
@@ -97,25 +106,29 @@ - (void)willInsertSections:(NSIndexSet *)sections
     }
     
     [self insertSections:sectionArray ofKind:kind atIndexSet:sections completion:nil];
-    [self batchLayoutNodes:nodes ofKind:kind atIndexPaths:_pendingIndexPaths[kind] completion:nil];
-    _pendingNodes[kind] = nil;
-    _pendingIndexPaths[kind] = nil;
+    [self batchLayoutNodes:nodes ofKind:kind atIndexPaths:_pendingIndexPaths[kind] completion:^(NSArray *nodes, NSArray *indexPaths) {
+      [self insertNodes:nodes ofKind:kind atIndexPaths:indexPaths completion:nil];
+    }];
+    [_pendingNodes removeObjectForKey:kind];
+    [_pendingIndexPaths removeObjectForKey:kind];
   }];
 }
 
 - (void)willDeleteSections:(NSIndexSet *)sections
 {
-  [[self supplementaryKinds] enumerateObjectsUsingBlock:^(NSString *kind, NSUInteger idx, BOOL *stop) {
+  NSArray *kinds = [self supplementaryKinds];
+  for (NSString *kind in kinds) {
     NSArray *indexPaths = ASIndexPathsForMultidimensionalArrayAtIndexSet([self editingNodesOfKind:kind], sections);
     
     [self deleteNodesOfKind:kind atIndexPaths:indexPaths completion:nil];
     [self deleteSectionsOfKind:kind atIndexSet:sections completion:nil];
-  }];
+  }
 }
 
 - (void)prepareForReloadSections:(NSIndexSet *)sections
 {
-  [[self supplementaryKinds] enumerateObjectsUsingBlock:^(NSString *kind, NSUInteger idx, BOOL *stop) {
+  NSArray *kinds = [self supplementaryKinds];
+  for (NSString *kind in kinds) {
```

Comment:
Now that I see you did the whole file - which is badass!! - it probably would be better to save the lines and do the self call inline with the for ().